### DISCUSSION TITLE
TEST: SAVE "tpi:cd dirname"

### DISCUSSION THREAD

**Discussion by factus10** [timex-sinclair-projects/TS-Pico#21] [General]
**Test the ability to change directories**

We have experienced and have reports that the CD command may cause the TS-Pico to hang. We'd like you to test it as extensively as you can.

### Testing the command

The CD command can move to different subdirectories one or several levels at a time. For example:
```
SAVE "tpi:cd a-m"
```
Using the SD card provided with your TS-Pico, the command above will move you to the A-M folder (FAT32 file systems, like the one on your SD card, are case insensitive but case preserving. They retain whatever case you used to create the file or directory name but don't care whether you use lower case, UPPER case, or MiXeD case to access them.)

The following command will jump directly from the top level to a directory two levels down:
```
SAVE "tpi:cd a-m/a"
```

Once you are in a directory, use **LOAD "tpi:dir"** to verify the successful change directory and the contents of the directory.

#### Recommended steps

1. If you have not downloaded the ZIP package for this GitHub, [please do so using this link](https://github.com/timex-sinclair-projects/TS-Pico/archive/refs/heads/Beta.zip). It contains a folder called _sdcard_ that has within it a ZIP file of the SD card image. You can use this to restore your SD card to it initial state.
2. Examine the file system on your SD card right now, using your Windows, Mac, or Linux computer. Drill down through all the folders. The folder and file names should be unremarkable.
3. Put the SD card back in your TS-Pico and power up your 2068.
4. Test variations of the CD command. Use the **LOAD "tpi:dir"** command to confirm the directory contents.
5. Attempt to mount (**LOAD "tpi:filename"**) a program within the folder. Attempt to load a file (**LOAD ""**) within the TAP file.
6. CD to another directory. Repeat step 5.

#### If the TS-Pico hangs

**Before you do anything else:** turn off your 2068, take out the SD card and examine it on your Windows, Mac, Linux computer. Look at all the directories, but especially the one you tried to CD to, for corruption. 

**Next:** try putting the SD card back in your TS-Pico, start your 2068 up and try to CD to the directory you tried when the Pico hung. 

**After that:** format your SD card and restore it from the contents of the SDcard.zip file.

**Report your testing and all results below**

**Comment by Sasounet** [comment]
Hi all,

did further test today. The only pattern I can see so far is that the problem occurs when two consecutive cd commands are issued. They do not need to be the first two commands after boot. If you type another command in between the two cd command, like for example LOAD "tpi:dir", it will not crash. I was also able to make it crash with a cd .. just after another cd something.

Again the problem is sporadic, about 1 time out of two. I made sure I was waiting long enough between the two CD command to make sure the heart beat was flashing as it should.

I check the SD card on my Mac, and everything looked fine. I ran the Disk Utility S.O.S program to analyse the file system and a bunch of warning were issued, most of them related to the crap the Mac automatically put on any disk when you mount them, but at least one tap file was had a warning message. I stupidly lost the log file of the S.O.S. utility so I cannot post it here.

 Francois

**Comment by rcalcagno** [comment]
Francois: did you wait until the LED resumes blinking, before issuing the two consecutive CD?

**Reply by Sasounet** [reply]
Yes I make sure of it. Anyway, I am a slow typist, especially on the bad TS-2068 keyboard, so the LED has resumed blinking for a while when I am ready to press ENTER. I can take a video of the problem if that could help.

An additional information: the verbose mode does not report the change of directory, the PICO crash before that.

Beside me and, I think, David, has anybody else been able to reproduce this ?

**Comment by rcalcagno** [comment]
Sorry Francois, I missed that you've been waiting for the heartbeat before issuing further commands. I will try to reproduce the issue, and if I can't will ask you for assistance to repeat the whole process.

**Comment by rcalcagno** [comment]
I found a possible culprit in line #1940; it was a break statement instead of continue. Attached is a small program that I had running a couple minutes, without failure. Feel free to try variations with different directories to see if you can make it fail:



![CD test prog](https://github.com/user-attachments/assets/c5cece6d-aae4-4257-ba41-0ffb3b9ca329)



https://github.com/user-attachments/assets/7f53bbf2-1885-4873-9fd1-d82c6cc339f8

**Comment by factus10** [comment]
Fingers crossed! 

Ricardo, if you can update the beta branch here, we can ask folks to download the entire GitHub as a zip file (I found the link last night).

**Comment by rcalcagno** [comment]
> Fingers crossed!
Indeed!
 
I already updated the file on /src/src_new_v1.1/TS/tspico.py please let me know if that's enough or should I put the file somewhere else

**Comment by factus10** [comment]
<img width="934" alt="Screenshot 2024-08-14 at 5 36 41 PM" src="https://github.com/user-attachments/assets/3a33dfce-54cc-4a3e-bb28-bfbf30490377">

1. Go to the [home page for this GitHub](https://github.com/timex-sinclair-projects/TS-Pico)
2. Select Beta
3. Click on the green Code button dropdown
4. Click on Download ZIP

**Reply by atrionfo** [reply]
I download this zip file and it worked fine.

**Comment by rcalcagno** [comment]
STOP THE PRESS!!! Please read comments on issue #15

**Comment by factus10** [comment]
Skreeeeeee crash! Press stopped.

**Comment by BetaTesterK** [comment]
Even though this looks like it's no longer needed, I'll still run the test.  I don't think all the kinks will be gone yet, based on the comments on issue #15.

Does look like progress, though.

**Comment by atrionfo** [comment]
I updated my TS-Pico using Thonny just a few minutes ago with the new main.py that hopefully fixes some of the problems with the CD command.  My TS-Pico booted after I pressed the TS Reset button on it.

Not every time, but nearly every time, I need to press the TS Reset when I turn on my T/S 2068, else I never get past a screen that is full of garbage.  I will avoid pressing the Pico Reset button again (I think Ricardo said, "Never press that button" during Wednesday's Timex Zoom meeting).  I'll also pay closer attention to the blinking of the LED on the Pico.

**Reply by factus10** [reply]
The scrambled screen is a function of how long it takes the 2068 to boot vs the Pico. Relatively speaking, the Pico is very slow.

Having said that, it sounds like boot time may have grown in this version. It's doing a little more, so it makes sense.

I'll start a bug issue for that.

**Comment by atrionfo** [comment]
I typed in most of Ricardo's BASIC program to change directories.  I've taken a picture of the screen and have attached it here.
![CD_Commands_BASIC_Cropped](https://github.com/user-attachments/assets/b12ede86-63ea-4e2b-b717-c251491e6f60)
I did not include the last line to GOTO 10

When I RUN the program, it takes a little while (20 seconds, according to my stopwatch).  It seems like the TS-Pico has crashed, and then I get a message that says, "J Invalid I/O device, 10:1"  When this happens, the LED on the Pico is blinking fast (an error message, I think).  When I try to get a directory listing (LOAD "tpi:dir"), after 20 seconds, I get and "J Invalid I/O device" message.  I tried other commands and I get the same reaction: After 20 seconds, I get an Invalid I/O device."

At this point, is my SD card still mounted to the TS-Pico?  Does turning off my TS-Pico corrupt the card?  Should I press one of the reset buttons on the TS-Pico?  I'm not sure, so I just proceeded, throwing caution to the wind; it's fun being a beta tester-- I mean it!

I turned off my T/S 2068 and then rebooted.  This worked okay and a directory listing does give the expected listing.  I DID put the new main.py onto the card earlier, but it's hard to confirm it.  When I copied it the to SD card, Thonny warned me that I would be replacing a file with one of a different size.  I ran a GetInfo on the TS-Pico and the firmware revision is 1.1, so I guess the python update with the SD card fix doesn't change the firmware revision.

To test loading a game from a TAP file, I changed to the A-M/F directory.  This worked fine.  Then I mounted the Frogger TAP.  This appeared to work okay.  Then I tried to run it with: LOAD ""  I was given the message: 10 REM No file mounted.  I ran a GetInfo and it tells me that the Mounted File is: /A-M/F/frogger.tap  I tried running Frogger again with the following: LOAD ""  This time I got an error message: "J Invalid I/O device, 10:1"  Then the LED on the TS-Pico stayed on without blinking; the LED was solidly green.  When I pressed Enter, I got the no file mounted message again and the LED remained on without blinking.  When I tried to issue any tpi commands I got the "J Invalid I/O device" error again.

I turned off my T/S 2068.  I removed the micro-SD card from the TS-Pico.  I put the SD card into my Windows 10 system.  I checked the filenames on the card, and those all seem okay.  I don't have a utility to run to try examining the SD card more closely.  Ricardo, can you recommend a utility that will do this for me?

I'm not sure how to proceed from here.  I would like to make 100% certain that the new main.py is okay.  Can the firmware revision of it be updated?

I'll be available for beta-testing on and off today and this weekend.  What should I test now to help you all out with this CD command issue?

**Comment by factus10** [comment]
Resolved in 1.1c
